# ![](https://ga-dash.s3.amazonaws.com/production/assets/logo-9f88ae6c9c3871690e33280fcf557f33.png) Pandas for Exploratory Data Analysis I 
by [@josephofiowa](https://twitter.com/josephofiowa)

Pandas is the most prominent Python library for exploratory data analysis (EDA). The functions Pandas supports are integral to understanding, formatting, and preparing our data. Formally, we use Pandas to investigate, wrangle, munge, and clean our data. Pandas is the Swiss Army Knife of data manipulation!


We'll have two coding-heavy sessions on Pandas. In this one, we'll use Pandas to:
 - Read in a dataset
 - Investigate a dataset's integrity
 - Filter, sort, and manipulate a DataFrame's series

## About the Dataset: Adventureworks Cycles

<img align="right" src="http://lh6.ggpht.com/_XjcDyZkJqHg/TPaaRcaysbI/AAAAAAAAAFo/b1U3q-qbTjY/AdventureWorks%20Logo%5B5%5D.png?imgmax=800">

For today's Pandas exercises, we will be using a dataset developed by Microsoft for training purposes in SQL server, known the [Adventureworks Cycles 2014OLTP Database](https://github.com/Microsoft/sql-server-samples/releases/tag/adventureworks). It is based on a fictitious company called Adventure Works Cycles (AWC), a multinational manufacturer and seller of bicycles and accessories. The company is based in Bothell, Washington, USA and has regional sales offices in several countries. We will be looking at a single table from this database, the Production.Product table, which outlines some of the products this company sells. 

A full data dictionary can be viewed [here](https://www.sqldatadictionary.com/AdventureWorks2014/).


Let's take a closer look at the Production.Product table [data dictionary](https://www.sqldatadictionary.com/AdventureWorks2014/Production.Product.html), which is a description of the fields (columns) in the table (the .csv file we will import below):
- **ProductID** - Primary key for Product records.
- **Name** - Name of the product.
- **ProductNumber** - Unique product identification number.
- **MakeFlag** - 0 = Product is purchased, 1 = Product is manufactured in-house.
- **FinishedGoodsFlag** - 0 = Product is not a salable item. 1 = Product is salable.
- **Color** - Product color.
- **SafetyStockLevel** - Minimum inventory quantity.
- **ReorderPoint** - Inventory level that triggers a purchase order or work order.
- **StandardCost** - Standard cost of the product.
- **ListPrice** - Selling price.
- **Size** - Product size.
- **SizeUnitMeasureCode** - Unit of measure for the Size column.
- **WeightUnitMeasureCode** - Unit of measure for the Weight column.
- **DaysToManufacture** - Number of days required to manufacture the product.
- **ProductLine** - R = Road, M = Mountain, T = Touring, S = Standard
- **Class** - H = High, M = Medium, L = Low
- **Style** - W = Womens, M = Mens, U = Universal
- **ProductSubcategoryID** - Product is a member of this product subcategory. Foreign key to ProductSubCategory.ProductSubCategoryID.
- **ProductModelID** - Product is a member of this product model. Foreign key to ProductModel.ProductModelID.
- **SellStartDate** - Date the product was available for sale.
- **SellEndDate** - Date the product was no longer available for sale.
- **DiscontinuedDate** - Date the product was discontinued.
- **rowguid** - ROWGUIDCOL number uniquely identifying the record. Used to support a merge replication sample.
- **ModifiedDate** - Date and time the record was last updated.


## Importing Pandas

To [import a library](https://docs.python.org/3/reference/import.html), we write `import` and the library name. For Pandas, is it common to name the library `pd` so that when we reference a function from the Pandas library, we only write `pd` to reference the aliased [namespace](https://docs.python.org/3/tutorial/classes.html#python-scopes-and-namespaces) -- not `pandas`.

In [1]:
import pandas as pd
%matplotlib inline

In [2]:
# we can see the details about the imported package by referencing its private class propertys:
print(f'I am using {pd.__name__} \
Version: {pd.__version__}.\n\
It is installed at: {pd.__path__}')

I am using pandas Version: 1.0.1.
It is installed at: ['/Users/jeffhale/miniconda3/envs/main/lib/python3.7/site-packages/pandas']


## Reading in Data

Pandas dramatically simplifies the process of reading in data. When we say "reading in data," we mean loading a file into our machine's memory.

When you have a CSV, for example, and then you double-click to open it in Microsoft Excel, the open file is "read into memory." You can now manipulate the CSV.

When we read data into memory in Python, we are creating an object. We will soon explore this object. _(And, as an aside, when we have a file that is greater than the size of our computer's memory, this is approaching "Big Data.")_

Because we are working with a CSV, we will use the [read CSV](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html) method.<br>A [delimiter](https://en.wikipedia.org/wiki/Delimiter-separated_values) is a character that separates fields (columns) in the imported file. Just because a file says `.csv` does not necessarily mean that a comma is used as the delimiter. In this case, we have a tab character as the delimiter for our columns, so we will be using `sep='\t'` to tell pandas to 'cut' the columns every time it sees a [tab character in the file](http://vim.wikia.com/wiki/Showing_the_ASCII_value_of_the_current_character).

In [3]:
# read the dataset as a DataFrame into a variable named 'prod'
prod = pd.read_csv('../data/Production.Product.csv', sep='\t')

In [4]:
# show the head of this dataset
prod.head()

ProductID                   Name ProductNumber  MakeFlag  \
0          1        Adjustable Race       AR-5381         0   
1          2           Bearing Ball       BA-8327         0   
2          3        BB Ball Bearing       BE-2349         1   
3          4  Headset Ball Bearings       BE-2908         0   
4        316                  Blade       BL-2036         1   

   FinishedGoodsFlag Color  SafetyStockLevel  ReorderPoint  StandardCost  \
0                  0   NaN              1000           750           0.0   
1                  0   NaN              1000           750           0.0   
2                  0   NaN               800           600           0.0   
3                  0   NaN               800           600           0.0   
4                  0   NaN               800           600           0.0   

   ListPrice  ... ProductLine Class Style  ProductSubcategoryID  \
0        0.0  ...         NaN   NaN   NaN                   NaN   
1        0.0  ...         NaN   NaN   NaN                   NaN   
2        0.0  ...         NaN   NaN   NaN                   NaN   
3        0.0  ...         NaN   NaN   NaN                   NaN   
4        0.0  ...         NaN   NaN   NaN                   NaN   

   ProductModelID        SellStartDate SellEndDate DiscontinuedDate  \
0             NaN  2008-04-30 00:00:00         NaN              NaN   
1             NaN  2008-04-30 00:00:00         NaN              NaN   
2             NaN  2008-04-30 00:00:00         NaN              NaN   
3             NaN  2008-04-30 00:00:00         NaN              NaN   
4             NaN  2008-04-30 00:00:00         NaN              NaN   

                                  rowguid                   ModifiedDate  
0  {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}  2014-02-08 10:01:36.827000000  
1  {58AE3C20-4F3A-4749-A7D4-D568806CC537}  2014-02-08 10:01:36.827000000  
2  {9C21AED2-5BFA-4F18-BCB8-F11638DC2E4E}  2014-02-08 10:01:36.827000000  
3  {ECFED6CB-51FF-49B5-B06C-7D8AC834DB8B}  2014-02-08 10:01:36.827000000  
4  {E73E9750-603B-4131-89F5-3DD15ED5FF80}  2014-02-08 10:01:36.827000000  

[5 rows x 25 columns]

*Documentation Pause*

How did we know how to use `pd.read_csv`? Let's take a look at the [documentation](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html). Note the first argument required (`filepath`).
> Take a moment to dissect other arguments and options when reading in data.

We have just created a data structure called a `DataFrame`. See?

In [5]:
# show the type of this 'prod' dataset
type(prod)

pandas.core.frame.DataFrame

## Inspecting our DataFrame: The basics

We'll now perform basic operations on the DataFrame, denoted with comments.

In [6]:
# print the first and last 3 rows
prod.head(3)

ProductID             Name ProductNumber  MakeFlag  FinishedGoodsFlag  \
0          1  Adjustable Race       AR-5381         0                  0   
1          2     Bearing Ball       BA-8327         0                  0   
2          3  BB Ball Bearing       BE-2349         1                  0   

  Color  SafetyStockLevel  ReorderPoint  StandardCost  ListPrice  ...  \
0   NaN              1000           750           0.0        0.0  ...   
1   NaN              1000           750           0.0        0.0  ...   
2   NaN               800           600           0.0        0.0  ...   

  ProductLine Class Style  ProductSubcategoryID  ProductModelID  \
0         NaN   NaN   NaN                   NaN             NaN   
1         NaN   NaN   NaN                   NaN             NaN   
2         NaN   NaN   NaN                   NaN             NaN   

         SellStartDate SellEndDate DiscontinuedDate  \
0  2008-04-30 00:00:00         NaN              NaN   
1  2008-04-30 00:00:00         NaN              NaN   
2  2008-04-30 00:00:00         NaN              NaN   

                                  rowguid                   ModifiedDate  
0  {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}  2014-02-08 10:01:36.827000000  
1  {58AE3C20-4F3A-4749-A7D4-D568806CC537}  2014-02-08 10:01:36.827000000  
2  {9C21AED2-5BFA-4F18-BCB8-F11638DC2E4E}  2014-02-08 10:01:36.827000000  

[3 rows x 25 columns]

In [7]:
prod.tail(3)

ProductID                Name ProductNumber  MakeFlag  FinishedGoodsFlag  \
501        997  Road-750 Black, 44    BK-R19B-44         1                  1   
502        998  Road-750 Black, 48    BK-R19B-48         1                  1   
503        999  Road-750 Black, 52    BK-R19B-52         1                  1   

     Color  SafetyStockLevel  ReorderPoint  StandardCost  ListPrice  ...  \
501  Black               100            75      343.6496     539.99  ...   
502  Black               100            75      343.6496     539.99  ...   
503  Black               100            75      343.6496     539.99  ...   

    ProductLine Class Style  ProductSubcategoryID  ProductModelID  \
501          R     L     U                    2.0            31.0   
502          R     L     U                    2.0            31.0   
503          R     L     U                    2.0            31.0   

           SellStartDate SellEndDate DiscontinuedDate  \
501  2013-05-30 00:00:00         NaN              NaN   
502  2013-05-30 00:00:00         NaN              NaN   
503  2013-05-30 00:00:00         NaN              NaN   

                                    rowguid                   ModifiedDate  
501  {44CE4802-409F-43AB-9B27-CA53421805BE}  2014-02-08 10:01:36.827000000  
502  {3DE9A212-1D49-40B6-B10A-F564D981DBDE}  2014-02-08 10:01:36.827000000  
503  {AE638923-2B67-4679-B90E-ABBAB17DCA31}  2014-02-08 10:01:36.827000000  

[3 rows x 25 columns]

Notice that `.head()` is a method (denoted by parantheses), so it takes arguments.

**Class Question:** 
- What do you think changes if we pass a different number `head()` argument?
- How would we print the last 5 rows?

In [8]:
# identify the shape (rows by columns)
prod.shape

(504, 25)

Here we have 504 rows, and 25 columns. This is a tuple, so we can extract the parts using indexing:

In [9]:
# print the number of rows

len(prod.index) #derek


504

In [10]:
prod.shape[0] #clay

504

In [11]:
rows, columns = prod.shape # Bibor
rows


504

# Anything in pandas confusing?

In [12]:
prod.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 504 entries, 0 to 503
Data columns (total 25 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   ProductID              504 non-null    int64  
 1   Name                   504 non-null    object 
 2   ProductNumber          504 non-null    object 
 3   MakeFlag               504 non-null    int64  
 4   FinishedGoodsFlag      504 non-null    int64  
 5   Color                  256 non-null    object 
 6   SafetyStockLevel       504 non-null    int64  
 7   ReorderPoint           504 non-null    int64  
 8   StandardCost           504 non-null    float64
 9   ListPrice              504 non-null    float64
 10  Size                   211 non-null    object 
 11  SizeUnitMeasureCode    176 non-null    object 
 12  WeightUnitMeasureCode  205 non-null    object 
 13  Weight                 205 non-null    float64
 14  DaysToManufacture      504 non-null    int64  
 15  Produc

## Using the index
An [index](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Index.html) is an immutable ndarray implementing an ordered, sliceable set. The basic object storing axis labels for all pandas objects. Think of it as a 'row address' for your data frame (table). It is best practice to explicitly set the index of your dataframe, as these are commonly used as [primary keys](https://en.wikipedia.org/wiki/Primary_key) which can be used to [join](https://www.w3schools.com/sql/sql_join.asp) your dataframe to other dataframes.

The dataframe can store different types (int, string, datetime), and when importing data will automatically assign a number to each row, starting at zero and counting up. You can overwrite this, which is what we are going to do.

In [13]:
# display the index as it sits (auto-generated upon import)

prod.index

RangeIndex(start=0, stop=504, step=1)

In [14]:
# note that our auto-generated index has no name

print(prod.index.name)

None


In [15]:
prod.index.name = 'Auto-genated index'
prod.index.name

'Auto-genated index'

In [16]:
prod.head()

ProductID                   Name ProductNumber  MakeFlag  \
Auto-genated index                                                             
0                           1        Adjustable Race       AR-5381         0   
1                           2           Bearing Ball       BA-8327         0   
2                           3        BB Ball Bearing       BE-2349         1   
3                           4  Headset Ball Bearings       BE-2908         0   
4                         316                  Blade       BL-2036         1   

                    FinishedGoodsFlag Color  SafetyStockLevel  ReorderPoint  \
Auto-genated index                                                            
0                                   0   NaN              1000           750   
1                                   0   NaN              1000           750   
2                                   0   NaN               800           600   
3                                   0   NaN               800           600   
4                                   0   NaN               800           600   

                    StandardCost  ListPrice  ... ProductLine Class Style  \
Auto-genated index                           ...                           
0                            0.0        0.0  ...         NaN   NaN   NaN   
1                            0.0        0.0  ...         NaN   NaN   NaN   
2                            0.0        0.0  ...         NaN   NaN   NaN   
3                            0.0        0.0  ...         NaN   NaN   NaN   
4                            0.0        0.0  ...         NaN   NaN   NaN   

                    ProductSubcategoryID  ProductModelID        SellStartDate  \
Auto-genated index                                                              
0                                    NaN             NaN  2008-04-30 00:00:00   
1                                    NaN             NaN  2008-04-30 00:00:00   
2                                    NaN             NaN  2008-04-30 00:00:00   
3                                    NaN             NaN  2008-04-30 00:00:00   
4                                    NaN             NaN  2008-04-30 00:00:00   

                   SellEndDate DiscontinuedDate  \
Auto-genated index                                
0                          NaN              NaN   
1                          NaN              NaN   
2                          NaN              NaN   
3                          NaN              NaN   
4                          NaN              NaN   

                                                   rowguid  \
Auto-genated index                                           
0                   {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}   
1                   {58AE3C20-4F3A-4749-A7D4-D568806CC537}   
2                   {9C21AED2-5BFA-4F18-BCB8-F11638DC2E4E}   
3                   {ECFED6CB-51FF-49B5-B06C-7D8AC834DB8B}   
4                   {E73E9750-603B-4131-89F5-3DD15ED5FF80}   

                                     ModifiedDate  
Auto-genated index                                 
0                   2014-02-08 10:01:36.827000000  
1                   2014-02-08 10:01:36.827000000  
2                   2014-02-08 10:01:36.827000000  
3                   2014-02-08 10:01:36.827000000  
4                   2014-02-08 10:01:36.827000000  

[5 rows x 25 columns]

In [17]:
# Here we are looking at two columns and an index;
# the one on the left is the index (automatically generated upon import by pandas)
# 'ProductID' is our PK (primary key) from our imported table. 'Name' is a data column.
# Notice that the generated index starts at zero and our PK starts at 1.

prod[['Name', 'ProductID',]].head(3)

Name  ProductID
Auto-genated index                            
0                   Adjustable Race          1
1                      Bearing Ball          2
2                   BB Ball Bearing          3

In [18]:
type(prod[['Name']])

pandas.core.frame.DataFrame

In [19]:
# Setting the index overwrites the automatically generated index
# with our PK, which resided in the 'ProductID' column.

prod = prod.set_index('ProductID')

In [20]:
prod.head()

Name ProductNumber  MakeFlag  FinishedGoodsFlag  \
ProductID                                                                     
1                Adjustable Race       AR-5381         0                  0   
2                   Bearing Ball       BA-8327         0                  0   
3                BB Ball Bearing       BE-2349         1                  0   
4          Headset Ball Bearings       BE-2908         0                  0   
316                        Blade       BL-2036         1                  0   

          Color  SafetyStockLevel  ReorderPoint  StandardCost  ListPrice Size  \
ProductID                                                                       
1           NaN              1000           750           0.0        0.0  NaN   
2           NaN              1000           750           0.0        0.0  NaN   
3           NaN               800           600           0.0        0.0  NaN   
4           NaN               800           600           0.0        0.0  NaN   
316         NaN               800           600           0.0        0.0  NaN   

           ... ProductLine Class  Style  ProductSubcategoryID ProductModelID  \
ProductID  ...                                                                 
1          ...         NaN   NaN    NaN                   NaN            NaN   
2          ...         NaN   NaN    NaN                   NaN            NaN   
3          ...         NaN   NaN    NaN                   NaN            NaN   
4          ...         NaN   NaN    NaN                   NaN            NaN   
316        ...         NaN   NaN    NaN                   NaN            NaN   

                 SellStartDate SellEndDate  DiscontinuedDate  \
ProductID                                                      
1          2008-04-30 00:00:00         NaN               NaN   
2          2008-04-30 00:00:00         NaN               NaN   
3          2008-04-30 00:00:00         NaN               NaN   
4          2008-04-30 00:00:00         NaN               NaN   
316        2008-04-30 00:00:00         NaN               NaN   

                                          rowguid  \
ProductID                                           
1          {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}   
2          {58AE3C20-4F3A-4749-A7D4-D568806CC537}   
3          {9C21AED2-5BFA-4F18-BCB8-F11638DC2E4E}   
4          {ECFED6CB-51FF-49B5-B06C-7D8AC834DB8B}   
316        {E73E9750-603B-4131-89F5-3DD15ED5FF80}   

                            ModifiedDate  
ProductID                                 
1          2014-02-08 10:01:36.827000000  
2          2014-02-08 10:01:36.827000000  
3          2014-02-08 10:01:36.827000000  
4          2014-02-08 10:01:36.827000000  
316        2014-02-08 10:01:36.827000000  

[5 rows x 24 columns]

In [21]:
# Note how our index property has changed as a result

In [22]:
# And our index has also inherited the name of our 'ProductID' column

## Column headers and datatypes

In [23]:
# print the columns
prod.columns

Index(['Name', 'ProductNumber', 'MakeFlag', 'FinishedGoodsFlag', 'Color',
       'SafetyStockLevel', 'ReorderPoint', 'StandardCost', 'ListPrice', 'Size',
       'SizeUnitMeasureCode', 'WeightUnitMeasureCode', 'Weight',
       'DaysToManufacture', 'ProductLine', 'Class', 'Style',
       'ProductSubcategoryID', 'ProductModelID', 'SellStartDate',
       'SellEndDate', 'DiscontinuedDate', 'rowguid', 'ModifiedDate'],
      dtype='object')

In [24]:
# cast into a list
list(prod.columns)

['Name',
 'ProductNumber',
 'MakeFlag',
 'FinishedGoodsFlag',
 'Color',
 'SafetyStockLevel',
 'ReorderPoint',
 'StandardCost',
 'ListPrice',
 'Size',
 'SizeUnitMeasureCode',
 'WeightUnitMeasureCode',
 'Weight',
 'DaysToManufacture',
 'ProductLine',
 'Class',
 'Style',
 'ProductSubcategoryID',
 'ProductModelID',
 'SellStartDate',
 'SellEndDate',
 'DiscontinuedDate',
 'rowguid',
 'ModifiedDate']

In [25]:
# examine the datatypes of the columns
# note that these were automatically inferred by pandas upon import!

prod.dtypes

Name                      object
ProductNumber             object
MakeFlag                   int64
FinishedGoodsFlag          int64
Color                     object
SafetyStockLevel           int64
ReorderPoint               int64
StandardCost             float64
ListPrice                float64
Size                      object
SizeUnitMeasureCode       object
WeightUnitMeasureCode     object
Weight                   float64
DaysToManufacture          int64
ProductLine               object
Class                     object
Style                     object
ProductSubcategoryID     float64
ProductModelID           float64
SellStartDate             object
SellEndDate               object
DiscontinuedDate         float64
rowguid                   object
ModifiedDate              object
dtype: object

In [26]:
type(prod.dtypes)

pandas.core.series.Series

**Class Question:** Why do datatypes matter? What operations could we perform on some datatypes that we could not on others? Note the importance of this in checking dataset integrity.

## Selecting a Column

We can select columns in two ways. Either we treat the column as an attribute of the DataFrame or we index the DataFrame for a specific element (in this case, the element is a column name).

In [27]:
# show the difference in data types by using single and double square brackets
type(prod['Name'])

pandas.core.series.Series

In [29]:
prod.head(2)

Name ProductNumber  MakeFlag  FinishedGoodsFlag Color  \
ProductID                                                                     
1          Adjustable Race       AR-5381         0                  0   NaN   
2             Bearing Ball       BA-8327         0                  0   NaN   

           SafetyStockLevel  ReorderPoint  StandardCost  ListPrice Size  ...  \
ProductID                                                                ...   
1                      1000           750           0.0        0.0  NaN  ...   
2                      1000           750           0.0        0.0  NaN  ...   

          ProductLine Class  Style  ProductSubcategoryID ProductModelID  \
ProductID                                                                 
1                 NaN   NaN    NaN                   NaN            NaN   
2                 NaN   NaN    NaN                   NaN            NaN   

                 SellStartDate SellEndDate  DiscontinuedDate  \
ProductID                                                      
1          2008-04-30 00:00:00         NaN               NaN   
2          2008-04-30 00:00:00         NaN               NaN   

                                          rowguid  \
ProductID                                           
1          {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}   
2          {58AE3C20-4F3A-4749-A7D4-D568806CC537}   

                            ModifiedDate  
ProductID                                 
1          2014-02-08 10:01:36.827000000  
2          2014-02-08 10:01:36.827000000  

[2 rows x 24 columns]

In [30]:
prod['Name']

ProductID
1            Adjustable Race
2               Bearing Ball
3            BB Ball Bearing
4      Headset Ball Bearings
316                    Blade
               ...          
995        ML Bottom Bracket
996        HL Bottom Bracket
997       Road-750 Black, 44
998       Road-750 Black, 48
999       Road-750 Black, 52
Name: Name, Length: 504, dtype: object

In [ ]:
# select the Name column only, returned as a Series object

In [31]:
prod[['Name']]

Name
ProductID                       
1                Adjustable Race
2                   Bearing Ball
3                BB Ball Bearing
4          Headset Ball Bearings
316                        Blade
...                          ...
995            ML Bottom Bracket
996            HL Bottom Bracket
997           Road-750 Black, 44
998           Road-750 Black, 48
999           Road-750 Black, 52

[504 rows x 1 columns]

In [32]:
type(prod[['Name']])

pandas.core.frame.DataFrame

In [ ]:
# select the Name column only, returned as a DataFrame object

In [ ]:
# selecting > 1 column (must use double brackets!)

**Class Question:** What if we wanted to select a column that has a space in it? Which method from the above two would we use? Why?

## Renaming Columns

Perhaps we want to rename our columns. There are a few options for doing this.

Renaming **specific** columns by using a dictionary:

In [35]:
# rename one or more columns with a dictionary. Note: inplace=False will return a new dataframe,
# but leave the original dataframe untouched. Change this to True to modify the original dataframe.

list(prod.columns)

['Name',
 'ProductNumber',
 'MakeFlag',
 'FinishedGoodsFlag',
 'Color',
 'SafetyStockLevel',
 'ReorderPoint',
 'StandardCost',
 'ListPrice',
 'Size',
 'SizeUnitMeasureCode',
 'WeightUnitMeasureCode',
 'Weight',
 'DaysToManufacture',
 'ProductLine',
 'Class',
 'Style',
 'ProductSubcategoryID',
 'ProductModelID',
 'SellStartDate',
 'SellEndDate',
 'DiscontinuedDate',
 'rowguid',
 'ModifiedDate']

In [39]:
prod.rename(columns={'Name': 'NewName', 'ListPrice': 'Price'}, inplace=True)
prod.head()

NewName ProductNumber  MakeFlag  FinishedGoodsFlag  \
ProductID                                                                     
1                Adjustable Race       AR-5381         0                  0   
2                   Bearing Ball       BA-8327         0                  0   
3                BB Ball Bearing       BE-2349         1                  0   
4          Headset Ball Bearings       BE-2908         0                  0   
316                        Blade       BL-2036         1                  0   

          Color  SafetyStockLevel  ReorderPoint  StandardCost  Price Size  \
ProductID                                                                   
1           NaN              1000           750           0.0    0.0  NaN   
2           NaN              1000           750           0.0    0.0  NaN   
3           NaN               800           600           0.0    0.0  NaN   
4           NaN               800           600           0.0    0.0  NaN   
316         NaN               800           600           0.0    0.0  NaN   

           ... ProductLine Class  Style  ProductSubcategoryID ProductModelID  \
ProductID  ...                                                                 
1          ...         NaN   NaN    NaN                   NaN            NaN   
2          ...         NaN   NaN    NaN                   NaN            NaN   
3          ...         NaN   NaN    NaN                   NaN            NaN   
4          ...         NaN   NaN    NaN                   NaN            NaN   
316        ...         NaN   NaN    NaN                   NaN            NaN   

                 SellStartDate SellEndDate  DiscontinuedDate  \
ProductID                                                      
1          2008-04-30 00:00:00         NaN               NaN   
2          2008-04-30 00:00:00         NaN               NaN   
3          2008-04-30 00:00:00         NaN               NaN   
4          2008-04-30 00:00:00         NaN               NaN   
316        2008-04-30 00:00:00         NaN               NaN   

                                          rowguid  \
ProductID                                           
1          {694215B7-08F7-4C0D-ACB1-D734BA44C0C8}   
2          {58AE3C20-4F3A-4749-A7D4-D568806CC537}   
3          {9C21AED2-5BFA-4F18-BCB8-F11638DC2E4E}   
4          {ECFED6CB-51FF-49B5-B06C-7D8AC834DB8B}   
316        {E73E9750-603B-4131-89F5-3DD15ED5FF80}   

                            ModifiedDate  
ProductID                                 
1          2014-02-08 10:01:36.827000000  
2          2014-02-08 10:01:36.827000000  
3          2014-02-08 10:01:36.827000000  
4          2014-02-08 10:01:36.827000000  
316        2014-02-08 10:01:36.827000000  

[5 rows x 24 columns]

Renaming **ALL** columns with a new list of column names.

Note that the `pd.DataFrame.columns` property can be cast to a `list` type. Originally, it's a `pd.core.indexes.base.Index` object:

In [37]:
print('My columns look like:\n', prod.columns)
print('\nAnd the type is:\n', type(prod.columns))

My columns look like:
 Index(['Name', 'ProductNumber', 'MakeFlag', 'FinishedGoodsFlag', 'Color',
       'SafetyStockLevel', 'ReorderPoint', 'StandardCost', 'ListPrice', 'Size',
       'SizeUnitMeasureCode', 'WeightUnitMeasureCode', 'Weight',
       'DaysToManufacture', 'ProductLine', 'Class', 'Style',
       'ProductSubcategoryID', 'ProductModelID', 'SellStartDate',
       'SellEndDate', 'DiscontinuedDate', 'rowguid', 'ModifiedDate'],
      dtype='object')

And the type is:
 <class 'pandas.core.indexes.base.Index'>


In [ ]:
prod.columns = []

We can explicitly cast these to a list object as such, by using the built-in `list()` function:

In [ ]:
# print('Now my columns look like:\n', list(prod.columns))
# print('\nAnd are of type:\n', type(list(prod.columns)))

We can place these columns into a variable, `cols`:

In [40]:
# declare a list of strings - these strings will become the new column names

cols = list(prod.columns)
cols

['NewName',
 'ProductNumber',
 'MakeFlag',
 'FinishedGoodsFlag',
 'Color',
 'SafetyStockLevel',
 'ReorderPoint',
 'StandardCost',
 'Price',
 'Size',
 'SizeUnitMeasureCode',
 'WeightUnitMeasureCode',
 'Weight',
 'DaysToManufacture',
 'ProductLine',
 'Class',
 'Style',
 'ProductSubcategoryID',
 'ProductModelID',
 'SellStartDate',
 'SellEndDate',
 'DiscontinuedDate',
 'rowguid',
 'ModifiedDate']

In [46]:
cols[0] = 'BetName'
cols

['BetName',
 'ProductNumber',
 'MakeFlag',
 'FinishedGoodsFlag',
 'Color',
 'SafetyStockLevel',
 'ReorderPoint',
 'StandardCost',
 'Price',
 'Size',
 'SizeUnitMeasureCode',
 'WeightUnitMeasureCode',
 'Weight',
 'DaysToManufacture',
 'ProductLine',
 'Class',
 'Style',
 'ProductSubcategoryID',
 'ProductModelID',
 'SellStartDate',
 'SellEndDate',
 'DiscontinuedDate',
 'rowguid',
 'ModifiedDate']

We can use list indexing to mutate the columns we want:

In [50]:
# overwrite one of the column names using list indexing
prod.columns = cols
prod

BetName ProductNumber  MakeFlag  FinishedGoodsFlag  \
ProductID                                                                     
1                Adjustable Race       AR-5381         0                  0   
2                   Bearing Ball       BA-8327         0                  0   
3                BB Ball Bearing       BE-2349         1                  0   
4          Headset Ball Bearings       BE-2908         0                  0   
316                        Blade       BL-2036         1                  0   
...                          ...           ...       ...                ...   
995            ML Bottom Bracket       BB-8107         1                  1   
996            HL Bottom Bracket       BB-9108         1                  1   
997           Road-750 Black, 44    BK-R19B-44         1                  1   
998           Road-750 Black, 48    BK-R19B-48         1                  1   
999           Road-750 Black, 52    BK-R19B-52         1                  1   

           Color  SafetyStockLevel  ReorderPoint  StandardCost   Price Size  \
ProductID                                                                     
1            NaN              1000           750        0.0000    0.00  NaN   
2            NaN              1000           750        0.0000    0.00  NaN   
3            NaN               800           600        0.0000    0.00  NaN   
4            NaN               800           600        0.0000    0.00  NaN   
316          NaN               800           600        0.0000    0.00  NaN   
...          ...               ...           ...           ...     ...  ...   
995          NaN               500           375       44.9506  101.24  NaN   
996          NaN               500           375       53.9416  121.49  NaN   
997        Black               100            75      343.6496  539.99   44   
998        Black               100            75      343.6496  539.99   48   
999        Black               100            75      343.6496  539.99   52   

           ... ProductLine Class  Style  ProductSubcategoryID ProductModelID  \
ProductID  ...                                                                 
1          ...         NaN   NaN    NaN                   NaN            NaN   
2          ...         NaN   NaN    NaN                   NaN            NaN   
3          ...         NaN   NaN    NaN                   NaN            NaN   
4          ...         NaN   NaN    NaN                   NaN            NaN   
316        ...         NaN   NaN    NaN                   NaN            NaN   
...        ...         ...   ...    ...                   ...            ...   
995        ...         NaN    M     NaN                   5.0           96.0   
996        ...         NaN    H     NaN                   5.0           97.0   
997        ...          R     L      U                    2.0           31.0   
998        ...          R     L      U                    2.0           31.0   
999        ...          R     L      U                    2.0           31.0   

                 SellStartDate SellEndDate  DiscontinuedDate  \
ProductID                                                      
1          2008-04-30 00:00:00         NaN               NaN   
2          2008-04-30 00:00:00         NaN               NaN   
3          2008-04-30 00:00:00         NaN               NaN   
4          2008-04-30 00:00:00         NaN               NaN   
316        2008-04-30 00:00:00         NaN               NaN   
...                        ...         ...               ...   
995        2013-05-30 00:00:00         NaN               NaN   
996        2013-05-30 00:00:00         NaN               NaN   
997        2013-05-30 00:00:00         NaN               NaN   
998        2013-05-30 00:00:00         NaN               NaN   
999        2013-05-30 00:00:00         NaN               NaN   

                                          rowguid  \
ProductID                                 

Finally, we can set the `pd.DataFrame.columns` property (this is a settable class property), to the new `cols` list, overwriting the existing columns header names:

In [ ]:
# Note that our first column name has changed from 'Name' to 'NewName'

In [ ]:
# display the final result of the 'NewName' column
# prod.head(3)[['NewName']]

## Common Column Operations

While this is non-comprehensive, these are a few key column-specific data checks.


**Descriptive statistics:**  the minimum, first quartile, median, third quartile, and maximum.

(And more! The mean too.)

Five Number Summary (all assumes numeric data):
- **Min:** The smallest value in the column
- **Max:** The largest value in the column
- **Quartile:** A quartile is one fourth of our data
    - **First quartile:** This is the bottom most 25 percent
    - **Median:** The middle value. (Line all values biggest to smallest - median is the middle!) Also the 50th percentile
    - **Third quartile:** This the the top 75 percentile of our data


![](https://www.mathsisfun.com/data/images/quartiles-a.svg)

In [51]:
# note - .describe() *default* only checks numeric datatypes
# show .describe() for the 'MakeFlag', 'SafetyStockLevel', 'StandardCost' fields.

prod.describe()

MakeFlag  FinishedGoodsFlag  SafetyStockLevel  ReorderPoint  \
count  504.000000         504.000000        504.000000    504.000000   
mean     0.474206           0.585317        535.150794    401.363095   
std      0.499830           0.493157        374.112954    280.584715   
min      0.000000           0.000000          4.000000      3.000000   
25%      0.000000           0.000000        100.000000     75.000000   
50%      0.000000           1.000000        500.000000    375.000000   
75%      1.000000           1.000000       1000.000000    750.000000   
max      1.000000           1.000000       1000.000000    750.000000   

       StandardCost        Price       Weight  DaysToManufacture  \
count    504.000000   504.000000   205.000000         504.000000   
mean     258.602961   438.666250    74.069220           1.103175   
std      461.632808   773.602843   182.166588           1.492616   
min        0.000000     0.000000     2.120000           0.000000   
25%        0.000000     0.000000     2.880000           0.000000   
50%       23.372200    49.990000    17.900000           1.000000   
75%      317.075825   564.990000    27.350000           1.000000   
max     2171.294200  3578.270000  1050.000000           4.000000   

       ProductSubcategoryID  ProductModelID  DiscontinuedDate  
count            295.000000      295.000000               0.0  
mean              12.294915       37.444068               NaN  
std                9.860135       34.025442               NaN  
min                1.000000        1.000000               NaN  
25%                2.000000       11.000000               NaN  
50%               12.000000       26.000000               NaN  
75%               17.000000       48.500000               NaN  
max               37.000000      128.000000               NaN

**Value Counts:** `pd.Series.value_counts()` count the occurrence of each value within our series.

In [52]:
# show the most popular product colors (aggregated by count, descending by default)

prod['Color'].value_counts()

Black           93
Silver          43
Red             38
Yellow          36
Blue            26
Multi            8
Silver/Black     7
White            4
Grey             1
Name: Color, dtype: int64

**Unique values:** Determine the number of distinct values within a given series.

In [55]:
# What are the unique colors for the products?
prod['Color'].unique()

array([nan, 'Black', 'Silver', 'Red', 'White', 'Blue', 'Multi', 'Yellow',
       'Grey', 'Silver/Black'], dtype=object)

In [56]:
# how many distinct colors are there?
prod['Color'].nunique()

9

In [57]:
# We can also include nulls with .nunique() as such:
prod['Color'].nunique(dropna=False)

10

# From here
## Filtering on a Single Condition

Filtering and sorting are key processes that allow us to drill into the 'nitty gritty' and cross sections of our dataset.

To filter, we use a process called **Boolean Filtering**, wherein we define a Boolean condition, and use that Boolean condition to filer on our DataFrame.

Recall: our given dataset has a column `Color`. Let's see if we can find all products that are `Black`. Let's take a look at the first 10 rows of the dataframe to see how it looks as-is:

In [ ]:
# show the first 10 rows of the 'Color' column

By applying a `boolean mask` to this dataframe, `== 'Black'`, we can get the following:

In [ ]:
# set the previous output to == Black

Now we can use that 'mask' from above, and apply it to our full dataframe. Every time we have a `True` in a row, we return the row. If we have a `False` in that row, we do not return it. The result is a dataframe that only has rows where `Color` is `Black`:

In [ ]:
# show the full prod DataFrame where Color is Black

Let's calculate the **average ListPrice** for the **salable products**.

> Think: What are the component parts of this problem?

In [ ]:
# First, we need to get salable items. 
# Use your data dictionary from the beginning of this lesson.

Now, we need to find average list price of those above items. Let's just get the 'ListPrice' column for starters.

To get the average of that column, just take `.mean()`

We can take a shortcut and just use `.describe()` here:

**Sneak peek**: Another handy trick is to use `.hist()` to get a distribution of a continuous variable - in this case, `ListPrice`. We'll cover this more in future lessons:

## Filtering on Multiple Conditions

Here, we will filter on _multiple conditions_. Before, we filtered on rows where Color was Black. We also filtered where FinishedGoodsFlag was equal to 1. Let's see what happens when we filter on *both* simultaneously. 

The format for multiple conditions is:

`df[ (df['col1'] == value1) & (df['col2'] == value2) ]`

Or, more simply:

`df[ (CONDITION 1) & (CONDITION 2) ]`

Which eventually may evaluate to something like:

`df[ True & False ]`

...on a row-by-row basis. If the end result is `False`, the row is omitted.

_Don't forget parentheses in your conditions!_ This is a common mistake.

In [ ]:
# Let's look at a table where Color is Black, AND FinishedGoodsFlag is 1

In [ ]:
# Here we have an example of a list price of greater than 50, 
# OR a product size that is not equal to 'XL'

## Sorting

We can sort one column of our DataFrame as well.

In [ ]:
# let's sort by standard cost, descending

This one is a little more advanced, but it demonstrates a few things:
- Conversion of a `numpy.ndarray` object (return type of `pd.Series.unique()`) into a `pd.Series` object
- `pd.Series.sort_values` with the `by=` kwarg omitted (if only one column is the operand, `by=` doesn't need specified
- Alphabetical sort of a string field, `ascending=True` means A->Z
- Inclusion of nulls, `NaN` in a string field (versus omission with a float/int as prior example)

## Independent Exercises

Do your best to complete the following prompts. Don't hesitate to look at code we wrote together!

Print the first 4 rows of the whole DataFrame.

In [ ]:
# your answer here

How many rows are in the dataframe? Return the answer as an int.

In [ ]:
# your answer here

How many columns? Retrun the answer as an int.

In [ ]:
# your answer here

How many different product lines are there?

In [ ]:
# your answer here

What are the values of these product lines?

In [ ]:
# your answer here

Do the number of values for the product lines match the number you have using `.nunique()`? Why or why not?

In [ ]:
# your answer here

Take the output from your previous answer (using `.unique()`). Select the label corresponding to the `Road` product line using list indexing notation. How many characters are in this string?

In [ ]:
# your answer here

Do you notice anything odd about this?

In [ ]:
# your answer here

How many products are there for the `Road` product line? Don't forget what you just worked on above! Return your answer as an int.

In [ ]:
# your answer here

How many products are there in the `Women's` `Mountain` category? Return your answer as an int. _Hint: Use the data dictionary above!_

In [ ]:
# your answer here

**Challenge:** What are the top 3 _most expensive list price_ product that are either in the `Women's` `Mountain` category, _OR_ `Silver` in `Color`? Return your answer as a DataFrame object, with the `ProductID` index, `NewName` relabeled as `Name`, and `ListPrice` columns. Perform the statement in one execution, and do not mutate the source DataFrame.

In [ ]:
# your answer here

## Recap

We covered a lot of ground! It's ok if this takes a while to gel.

```python

# basic DataFrame operations
df.head()
df.tail()
df.shape
df.columns
df.Index

# selecting columns
df.column_name
df['column_name']

# renaming columns
df.rename({'old_name':'new_name'}, inplace=True)
df.columns = ['new_column_a', 'new_column_b']

# notable columns operations
df.describe() # five number summary
df['col1'].nunique() # number of unique values
df['col1'].value_counts() # number of occurrences of each value in column

# filtering
df[ df['col1'] < 50 ] # filter column to be less than 50
df[ (df['col1'] == value1) & (df['col2'] > value2) ] # filter column where col1 is equal to value1 AND col2 is greater to value 2

# sorting
df.sort_values(by='column_name', ascending = False) # sort biggest to smallest

```


It's common to refer back to your own code *all the time.* Don't hesistate to reference this guide! 🐼


